### PR TITLE
Added landing page.

### PR DESCRIPTION
**Pull request by dsjen**
- Separated footer and sock/feedback.
- Root url will redirect to /landing.

This is an early PR that adds a landing page to Insights.  The cover image is TBD, and this PR will be updated when that is available.

Example with interim cover image:
landing-page

@rocha @clintonb

### CODE REVIEW

File: acceptance_tests/test_landing.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+from acceptance_tests.mixins import FooterLegalMixin
+from acceptance_tests import OPEN_SOURCE_URL, RESEARCH_URL, SUPPORT_URL
+from pages import LandingPage
+
+
+_multiprocess_can_split_ = True
+
+
+class LandingTests(FooterLegalMixin, WebAppTest):
```
Comment (clintonb) [inline acceptance_tests/test_landing.py]: Testing the content of the page is nice; but, you should be testing the actually method of arriving at the page (e.g. if a user is not signed in the browser shows the landing page). I don't see that being tested.

File: analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    <body class="landing">
+        <div id="wrap-content-push-footer">
+            <header class="navbar navbar-default" role="banner" aria-label="main"
+                    id="nav">
+                <div class="container">
+                    <div class="navbar-header">
+                        {% include 'brand.html' with class_prefix='navbar' brand_link='/' %}
+                    </div>
+                    <div class="login-button-container">
+                        <a href="{% url 'courses:index' %}"><button type="button" class="btn btn-default pull-right">{% blocktrans %}Login{% endblocktrans %}</button></a>
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: Just use `{% trans %}` here to avoid potential issues with whitespace in the future.

Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: Also, this button should point to the login view (`{% url 'login' %}`). Django will handle redirection to the courses page.

File: analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                    </div>
+                </div>
+                <!-- / container -->
+
+                <div class="row masthead">
+                    <div class="cover-container">
+                        <img class="cover" src="{% static 'images/landing-map.png' %}"/>
+                    </div>
+                    <div class="col-xs-12 masthead-heading masthead-brand">
+                        <p>{% blocktrans %}Insights to help course teams improve courses.{% endblocktrans %}</p>
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: Replace Insights with `{{ application_name }}`.

**Comment by clintonb** [comment]
@shnayder do we anticipate that open source users will use this page as-is? The copy seems very edX-specific.

File: analytics_dashboard/analytics_dashboard/urls.py [hunk 4]
```diff
@@ -17,7 +17,7 @@
 
 urlpatterns = patterns(
     '',
-    url(r'^$', RedirectView.as_view(url=reverse_lazy('courses:index')), name='home'),
+    url(r'^$', RedirectView.as_view(url=reverse_lazy('landing')), name='home'),
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/urls.py]: This is not the desired behavior. As written here, if a user is logged in and visits the root page, he will end up on the static landing page. Since the user is logged in he should see the courses page. Instead of doing redirects here, point this to your `LandingView`. In the `dispute` method, check to see if the user is logged in and redirect there.

File: analytics_dashboard/templates/brand.html [hunk 5]
```diff
@@ -2,7 +2,12 @@
 {% load i18n %}
 {% load dashboard_extras %}
 
-<a class="{{ class_prefix }}-brand" href="/">
+{# set a default value for brand_link #}
```
Comment (clintonb) [inline analytics_dashboard/templates/brand.html]: This should not be done. See my comment in `urls.py`.

**Comment by clintonb** [comment]
Everything looks fine; but, I would like to resolve the redirect issue before this is merged.

**Comment by clintonb** [comment]
Rebase on master to resolve the build issue.

File: acceptance_tests/test_landing.py (hunk 1 above)
Comment (dsjen) [inline acceptance_tests/test_landing.py]: Good call--added tests for logged in user and logged out users.

File: analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html (hunk 2 above)
Comment (dsjen) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: Updated.

File: analytics_dashboard/analytics_dashboard/urls.py (hunk 4 above)
Comment (dsjen) [inline analytics_dashboard/analytics_dashboard/urls.py]: Check for authentication state and redirect as necessary in the `dispatch` method. :)

File: analytics_dashboard/templates/brand.html (hunk 5 above)
Comment (dsjen) [inline analytics_dashboard/templates/brand.html]: Updated.

**Comment by shnayder** [comment]
@clintonb great question re: OS. We should have thought that through better. Short term steps:
- The logo should use the same theming pattern as in the rest of the app
- Let's have a way to turn off the research blurb
- Make sure the support center link is template-ized
- The rest of the text seems generic enough for now.

Long term steps:
- let's schedule a chat with Brian T and make a general theming plan going forward. We should hopefully be able to adapt the patterns we're planning for the rest of the platform.

**Comment by dsjen** [comment]
@shnayder, I'll need to make some adjustments for the research blurb to be configured.  The the logo is themed like the rest of the app and the support center link is the same link as the other support links.

**Comment by dsjen** [comment]
cover-image

@shnayder, @marcotuts -- updated cover image

File: acceptance_tests/__init__.py [hunk 6]
```diff
@@ -14,6 +14,9 @@ def str2bool(s):
 PLATFORM_NAME = os.environ.get('PLATFORM_NAME', 'edX')
 APPLICATION_NAME = os.environ.get('APPLICATION_NAME', 'Insights')
 SUPPORT_URL = os.environ.get('SUPPORT_URL', 'http://example.com/')
+OPEN_SOURCE_URL = os.environ.get('OPEN_SOURCE_URL', 'http://example.com/')
+RESEARCH_URL = os.environ.get('RESEARCH_URL', 'http://example.com/')
+SHOW_LANDING_RESEARCH = os.environ.get('SHOW_LANDING_RESEARCH', True)
```
Comment (clintonb) [inline acceptance_tests/__init__.py]: Use `str2bool`.

File: acceptance_tests/mixins.py [hunk 7]
```diff
@@ -148,6 +158,13 @@ def login_with_lms(self):
         self.lms_login_page.login(LMS_USERNAME, LMS_PASSWORD)
 
 
+class LogoutMixin(object):
+
+    def logout(self):
```
Comment (clintonb) [inline acceptance_tests/mixins.py:168]: This doesn't really need to be a mixin since it's only used at one location.

File: acceptance_tests/__init__.py (hunk 6 above)
Comment (dsjen) [inline acceptance_tests/__init__.py]: What package is `str2bool` part of?

File: acceptance_tests/test_landing.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+_multiprocess_can_split_ = False
+
+
+class LandingTests(LoginMixin, LogoutMixin, FooterLegalMixin, WebAppTest):
+    def setUp(self):
+        super(LandingTests, self).setUp()
+        self.page = LandingPage(self.browser)
+
+    def test_page(self):
```
Comment (clintonb) [inline acceptance_tests/test_landing.py:16]: Call the parent method; otherwise, your footer tests will never run.

File: acceptance_tests/__init__.py (hunk 6 above)
Comment (clintonb) [inline acceptance_tests/__init__.py]: It's a method at the top of the file.

File: acceptance_tests/mixins.py (hunk 7 above)
Comment (dsjen) [inline acceptance_tests/mixins.py:168]: It just seemed fit nicely here close to the `LoginMixin` in case anyone was looking for logout at some point.

File: acceptance_tests/__init__.py (hunk 6 above)
Comment (dsjen) [inline acceptance_tests/__init__.py]: I see--thanks.

File: acceptance_tests/test_landing.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        # landing page will not be viewable by logged in users
+        self.page.browser.get(self.page.path)
+        self.assertTrue(not self.page.is_browser_on_page())
```
Comment (clintonb) [inline acceptance_tests/test_landing.py]: `assertFalse`

File: acceptance_tests/test_landing.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+        # landing page only accessible to logged out users
+        self.logout()
+        # logout page will redirect to the landing page
+        self._test_entire_page()
```
Comment (clintonb) [inline acceptance_tests/test_landing.py]: It's not necessary to test the page twice. Simply verify that the browser is on the correct page after logout.

File: acceptance_tests/test_landing.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+        # landing page is visible at the root level
+        self.page.visit()
+        self._test_entire_page()
```
Comment (clintonb) [inline acceptance_tests/test_landing.py]: Making these two calls a method doesn't save much of anything. I would just call `_test_lenses` and `_test_audience_messages` directly.

File: acceptance_tests/test_landing.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+    def _test_entire_page(self):
+        self._test_lenses()
+        self._test_audience_messages()
+
+    def _test_lenses(self):
+        question_elements = self.page.q(css='.lens-question')
+        self.assertTrue(question_elements.present)
+
+        num_lenses = 3
```
Comment (clintonb) [inline acceptance_tests/test_landing.py]: Just use `len(expected_questions)` instead of hard-coding the value.

File: acceptance_tests/test_landing.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        self.assertEqual(len(question_elements), num_lenses)
```
Comment (clintonb) [inline acceptance_tests/test_landing.py]: This assertion isn't really necessary since you are checking the content below. If the number of elements is incorrect, the issue will be found when checking the content.

File: acceptance_tests/test_landing.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+        expected_questions = ['Who are my students?', 'What are students engaging with in my course?',
+                              'How well is my content supporting student learning?']
+        for i in range(num_lenses):
+            self.assertEqual(question_elements[i].text, expected_questions[i])
+
+        summary_elements = self.page.q(css='.lens-summary')
+        self.assertTrue(summary_elements.present)
```
Comment (clintonb) [inline acceptance_tests/test_landing.py:41]: Not necessary since you are checking the length below.

File: acceptance_tests/test_landing.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+        self.assertTrue(len(summary_elements), num_lenses)
+
+        lens_icon_elements = self.page.q(css='.lens-summary h1 i')
+        self.assertTrue(lens_icon_elements.present)
```
Comment (clintonb) [inline acceptance_tests/test_landing.py:45]: Not necessary since you are checking the length below.

File: acceptance_tests/test_landing.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+        # make sure that the icons are hidden from screen readers
+        for i in range(num_lenses):
+            self.assertEqual(lens_icon_elements.attrs('aria-hidden')[i], 'true')
+
+    def _test_audience_messages(self):
+        element = self.page.q(css='.audience-message')
+        self.assertTrue(element.present)
+
+        num_actions = 2
```
Comment (clintonb) [inline acceptance_tests/test_landing.py]: Use `len(expected_headers)`.

File: analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+
+
+{% block body %}
+    <body class="landing">
+        <div id="wrap-content-push-footer">
+            <header class="navbar navbar-default" role="banner" aria-label="main"
+                    id="nav">
+                <div class="container">
+                    <div class="navbar-header">
+                        {% include 'brand.html' with class_prefix='navbar' brand_link='/' %}
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: `brand_link` doesn't exist in `brand.html`.

File: analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+
+
+{% block title %}{% settings_value 'FULL_APPLICATION_NAME' %}{% endblock title %}
+
+
+{% block body %}
+    <body class="landing">
+        <div id="wrap-content-push-footer">
+            <header class="navbar navbar-default" role="banner" aria-label="main"
+                    id="nav">
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: Move to previous line.

File: analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+                <div class="row audience-message">
+                    {% captureas link_end %}</a>{% endcaptureas %}
+
+                    {% trans "Join the Open Source Community" as heading %}
+                    {% captureas link_start %}
+                        <a href="{{research_url}}" target="_blank">
+                    {% endcaptureas %}
+                    {% captureas description %}
+                        {% blocktrans %}
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: Add `trimmed` to trim whitespace.

File: analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+                    {% endcaptureas %}
+                    {% include "audience_message.html" with heading=heading description=description num_columns=num_audience_message_columns %}
+
+                    {% if show_research %}
+                        {% trans "Research at edX" as heading %}
+                        {% captureas link_start %}
+                            <a href="{{open_source_url}}" target="_blank">
+                        {% endcaptureas %}
+                        {% captureas description %}
+                            {% blocktrans %}
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: Add `trimmed` to trim whitespace.

File: analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+                        {% endcaptureas %}
+                        {% include "audience_message.html" with heading=heading description=description num_columns=num_audience_message_columns %}
+                    {% endif %}
+
+                    {% trans "Need Help?" as heading %}
+                    {% captureas link_start %}
+                        <a href="{{support_url}}" target="_blank">
+                    {% endcaptureas %}
+                    {% captureas description %}
+                        {% blocktrans %}
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: Add `trimmed` to trim whitespace.

File: analytics_dashboard/analytics_dashboard/views.py [hunk 22]
```diff
@@ -85,10 +85,10 @@ def get(self, request, *_args, **_kwargs):
         user = authenticate(username=username, password=password)
         login(request, user)
 
-        return redirect('/')
+        return redirect('/courses')
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/views.py]: Revert this change. The landing view will handle redirection.

File: analytics_dashboard/analytics_dashboard/views.py [hunk 23]
```diff
@@ -108,3 +108,29 @@ def logout_then_login(request, login_url=reverse_lazy('login'), current_app=None
...
+            return super(LandingView, self).dispatch(request, *args, **kwargs)
+        else:
+            return redirect('courses:index')
+
+    def get_context_data(self, **kwargs):
+        context = super(LandingView, self).get_context_data(**kwargs)
+        context['research_url'] = settings.RESEARCH_URL
+        context['open_source_url'] = settings.OPEN_SOURCE_URL
+        context['show_research'] = settings.SHOW_LANDING_RESEARCH and settings.RESEARCH_URL
+        num_audience_message_columns = 6
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/views.py]: This variable name is confusing. `audience_message_col_width` would be much less-confusing.

File: acceptance_tests/test_landing.py (hunk 12 above)
Comment (dsjen) [inline acceptance_tests/test_landing.py]: By being explicit about the number of lenses, we check for the case that there are more than 3 in `question_elements`.

File: acceptance_tests/test_landing.py (hunk 8 above)
Comment (dsjen) [inline acceptance_tests/test_landing.py:16]: Nice catch!

File: acceptance_tests/test_landing.py (hunk 9 above)
Comment (dsjen) [inline acceptance_tests/test_landing.py]: Done.

File: acceptance_tests/test_landing.py (hunk 10 above)
Comment (dsjen) [inline acceptance_tests/test_landing.py]: Makes sense.

File: acceptance_tests/test_landing.py (hunk 11 above)
Comment (dsjen) [inline acceptance_tests/test_landing.py]: Done.

File: analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html (hunk 19 above)
Comment (dsjen) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: Done.

File: analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html (hunk 20 above)
Comment (dsjen) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: Done.

File: analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html (hunk 21 above)
Comment (dsjen) [inline analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/landing.html]: Done.

File: gulpfile.js [hunk 24]
```diff
@@ -19,6 +19,7 @@
                 'analytics_dashboard/static/js/test/**/*.js'
             ],
             templates: [
+                'analytics_dashboard/analytics_dashboard/templates/analytics_dashboard/*.html',
```
Comment (clintonb) [inline gulpfile.js:22]: This path doesn't exist.

File: acceptance_tests/test_landing.py (hunk 12 above)
Comment (clintonb) [inline acceptance_tests/test_landing.py]: You still can; but, you don't need to hardcode the value when it can be pretty easily calculated.

File: gulpfile.js (hunk 24 above)
Comment (dsjen) [inline gulpfile.js:22]: This is where `landing.html` lives.

File: analytics_dashboard/analytics_dashboard/views.py (hunk 22 above)
Comment (dsjen) [inline analytics_dashboard/analytics_dashboard/views.py]: Done.

File: analytics_dashboard/analytics_dashboard/views.py (hunk 23 above)
Comment (dsjen) [inline analytics_dashboard/analytics_dashboard/views.py]: Updated to `audience_message_column_width`.

File: acceptance_tests/test_landing.py (hunk 14 above)
Comment (dsjen) [inline acceptance_tests/test_landing.py:41]: It's convenient to see where the error is first.  I see your point though.

File: analytics_dashboard/analytics_dashboard/urls.py [hunk 25]
```diff
@@ -18,7 +18,7 @@
 
 urlpatterns = patterns(
     '',
-    url(r'^$', RedirectView.as_view(url=reverse_lazy('courses:index')), name='home'),
+    url(r'^$', views.LandingView.as_view(), name='landing'),
```
Comment (clintonb) [inline analytics_dashboard/analytics_dashboard/urls.py:21]: Minor: Home would more accurately describe this view since it doesn't always point to the landing page.

**Comment by marcotuts** [comment]
hurray! thanks @dsjen :+1:

**Comment by shnayder** [comment]
:ship: :rocket: :cookie: :sunglasses:

**Comment by clintonb** [comment]
:+1:

File: acceptance_tests/mixins.py [hunk 26]
```diff
@@ -75,35 +75,50 @@ def assertTable(self, table_selector, columns, download_selector):
         self.assertValidHref(download_selector)
 
 
+class PageTestMixin(object):
+    def test_page(self):
+        pass
```
Comment (rocha) [inline acceptance_tests/mixins.py:80]: What is this mixin for? If it is just for defining an interface, then it may be unnecessary.

File: analytics_dashboard/analytics_dashboard/views.py [hunk 27]
```diff
@@ -108,3 +108,29 @@ def logout_then_login(request, login_url=reverse_lazy('login'), current_app=None
...
+
+class LandingView(TemplateView):
+    """
+    Displays a public landing page when users first come to the site.
+    """
+    template_name = "analytics_dashboard/landing.html"
+
+    def dispatch(self, request, *args, **kwargs):
+        """ Non logged in users will be directed to the landing page. """
+        if request.user.is_anonymous():
```
Comment (rocha) [inline analytics_dashboard/analytics_dashboard/views.py:121]: So that is how it is done! A lot better of the other things we tried. :smile:

**Comment by rocha** [comment]
:ship: :it: 

Nothing big on my end. The empty test mixin class is a bit weird python wise, but not a blocker.
